mixpanel_oauth_to_redshift.go

```diff
@@ -0,0 +1,179 @@
+)
+
+type MixpanelExport struct {
+	endpoint  string
+	version   string
+	apiKey    string
+	apiSecret string
+}
+
+func NewMixpanelExport() *MixpanelExport {
+	m := MixpanelExport{"https://data.mixpanel.com/api", "2.0", mixpanelAPIKey, mixpanelAPISecret}
+	return &m
+}
+
+// json.Marshal wraps a string with quotes which breaks URL params in API calls. This function fixes
+// that problem.
+func StringOrJsonMarshal(value interface{}) (string, error) {
+	strValue, ok := value.(string)
+	if !ok {
+		jsonValue, err := json.Marshal(value)
+		if err != nil {
+			return "", err
+		}
+		strValue = string(jsonValue)
+	}
+	return strValue, nil
+}
+
+func (m *MixpanelExport) ComputeSig(params map[string]interface{}) (string, error) {
+	var keys []string
+
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	joinedArgs := ""
+	for _, k := range keys {
+		joinedArgs += k
+		value, err := StringOrJsonMarshal(params[k])
+		if err != nil {
+			return "", err
+		}
+		joinedArgs += fmt.Sprintf("=%s", value)
+	}
+
+	hash := md5.Sum([]byte(joinedArgs + m.apiSecret))
+	return hex.EncodeToString(hash[:]), nil
+}
+
+func (m *MixpanelExport) Request(method string, params map[string]interface{}) ([]byte, error) {
+	params["api_key"] = m.apiKey
+	expireTime := time.Now().Add(*mixpanelExportRequestTimeout)
+	params["expire"] = expireTime.Unix()
+	params["format"] = "json"
+	// Delete original signature from the params if already set and assign the correct signature
+	// based on the new param values.
+	delete(params, "sig")
+	sig, err := m.ComputeSig(params)
+	if err != nil {
+		return nil, err
+	}
+	params["sig"] = sig
+
+	values := url.Values{}
+	for key, value := range params {
+		strValue, err := StringOrJsonMarshal(value)
+		if err != nil {
+			return nil, err
+		}
+		values.Set(key, strValue)
+	}
+	url := fmt.Sprintf("%s/%s/%s/?%s", m.endpoint, m.version, method, values.Encode())
+	log.Println(url)
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	return ioutil.ReadAll(resp.Body)
+}
+
+func CopyJsonDataFromS3ToRedshift(
+	db *sql.DB, redshiftTable string, exportFile string, awsAccessKeyId string,
+	awsSecretAccessKey string, jsonpathsFile string, awsRegion string) error {
+	copyCmd := fmt.Sprint(
+		"COPY ", redshiftTable, " FROM '", exportFile, "' WITH ",
+		" json '", jsonpathsFile,
+		"' region '", awsRegion,
+		"' timeformat 'epochsecs'")
+	log.Print("Executing command: ", copyCmd)
+	copyCmd += fmt.Sprintf(" credentials '%s=%s;%s=%s'", "aws_access_key_id", awsAccessKeyId, "aws_secret_access_key", awsSecretAccessKey)
+	_, err := db.Exec(copyCmd)
+	return err
+}
+
+func main() {
+	flag.Parse()
+	exportFile := fmt.Sprintf("%s/%s", *mixpanelExportDir, *mixpanelExportDate)
+
+	if *exportFromMixpanel {
+		mixpanelExport := NewMixpanelExport()
+		log.Println("Exporting mixpanel data for", *mixpanelExportDate)
+		params := map[string]interface{}{
+			"event":     [1]string{"authorize"},
```

Comment:
Is this the only thing that's schema specific? Might we parametrize this so we can easily pull other mixpanel events, instead of hard-coding?